libs/ml/include/ml/ops/ops.hpp

```diff
@@ -73,8 +97,93 @@ class Ops
     return is_training_;
   }
 
+  /// Calculates charge amount that is necessary to perform this Op on the input data Tensor(s)
+  /// with given shape(s).
+  virtual fetch::vm::ChargeAmount OpForwardCost(ShapeVector const &input_shapes)
+  {
+    FETCH_UNUSED(input_shapes);
+    FETCH_LOG_WARN("Ops", " not-implemented OpForwardCost() called! returned 0.");
+    return 0;  // TODO(VH): make me a pure virtual.
+  }
+
+  virtual void SetSliceOutputShape(Shape const &new_shape)
+  {
+    slice_output_shape_ = new_shape;
+  }
+
+  virtual void SetExpectedSliceInputShapes(ShapeVector const &new_shapes)
+  {
+    expected_slice_input_shapes_ = new_shapes;
+  }
+
+  /**
+   * @brief SliceOutputShape returns an output shape of the layer, if a singluar slice of an input
+   * data is given (e.g. batch size == 1)
+   */
+  virtual Shape const &SliceOutputShape() const
+  {
+    return slice_output_shape_;
+  }
+
+  /**
+   * @brief ExpectedSliceInputShapes returns a vector of shapes, that describes expected input
+   * slice shapes (e.g. when batch size of input data is 1)
+   */
+  virtual ShapeVector const &ExpectedSliceInputShapes() const
+  {
+    return expected_slice_input_shapes_;
+  }
+
+  /// OOP polymorphic wrapper around each Ops/Layer OpCode() static method.
+  virtual OpType OperationType() const  // TODO(VH): make a pure virtual.
+  {
+    return OpType::NONE;
+  }
+
 protected:
   bool is_training_ = true;
+
+  // TODO(VH): impl. filling it on compilation.
+  ShapeVector expected_slice_input_shapes_{};
+  Shape       slice_output_shape_{};
+  // TODO(VH): ^^ impl. serialisation of new fields.
+
+  std::string OutputShapeAsString()
+  {
+    std::vector<SizeType> const out_shape = this->SliceOutputShape();
+    std::stringstream           ss;
+    ss << " (out ";
+    if (out_shape.empty())
+    {
+      ss << "[??] )";
+      return ss.str();
+    }
+    ss << "[";
+    for (auto const &dim : out_shape)
+    {
+      ss << " " << dim;
+    }
+    ss << " ])";
+    return ss.str();
+  }
+
```

Comment:
I think we should consider making this a static free function utility that takes the op as input parameter? Or do we rely on private member access?

Reply:
we rely on private out_shape in this implementation, but yeap it could be extracted as a free function. However, I would separate it as a task "implement .summary() method for VMModel" or alike. This method is a debug/demo dummy to see what's happening while shapes are being linked and estimation is computed, and I believe in a production version we do not need it at all to prevent log flooding with unnecessary info.